### ISSUE TITLE
Device Registration with Certificate based authentication support

### ISSUE THREAD

**Issue by pratikmhatre11** [eclipse-hono/hono#3590]
I am using hono for one of my project. I have deployed MQTT-adapter, Registry-Server and Kafka through helm. 

Currently, I am authenticating the devices using username and password. I am able to create the new tenant and able to register multiple devices into the tenant. I am also able to send the data on MQTT telemetry topic and able to consume it through Kafka in the java backend application.

**My expected end Goal:**
Currently, I am authenticating the devices using username and password. Username-Password based authentication is working good for me. Now I want to avail the certificate based authentication.

**What I have done till now to implement certificate based authentication:**

I created a tenant named “DEFAULT_TENANT” using below steps:

```
Step-1: I have the certificates in the same folder where I am running this command
CERT=$(openssl x509 -in default_tenant-cert.pem -noout -pubkey | sed '/^---/d' | awk '{ printf "%s", $0 }')

Step-2:
TENANT_ID=DEFAULT_TENANT
REGISTRY_IP={{My Registry IP Here}}
curl --location 'https://${REGISTRY_IP}:28443/v1/tenants/${TENANT_ID}'
--header 'content-type: application/json'
--data '{
  "trusted-ca": [
    {
      "cert": "'${CERT}'"
    }
  ],
  "ext": {
    "messaging-type": "kafka"
  }
}'
```

After following above steps to check the status of tenant creation:

```
{{My Registry IP Here}}/v1/tenants/DEFAULT_TENANT

I get the below response:
{
    "ext": {
        "messaging-type": "kafka"
    },
    "trusted-ca": [
        {
            "id": "78d55b66-aad1-4250-971a-a20a8a8446dc",
            "subject-dn": "CN=DEFAULT_TENANT_CA,OU=Hono,O=Eclipse IoT,L=Ottawa,C=CA",
            "public-key": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEdVSIdH4aebiwLpVs5Zid3RhcnfevM6Bk369eiL6o8VNEBnb21oFcqynKSQFc3gcy8vbq/jpXg3Yi+DsDC23EeA==",
            "algorithm": "EC",
            "not-before": "2023-12-01T19:23:58Z",
            "not-after": "2024-11-30T19:23:58Z",
            "auto-provisioning-enabled": false,
            "auto-provision-as-gateway": false
        }
    ]
}
```

**Where I got stuck:**

I got stuck at device registration with certificate support. I have created the certificates for device “4711” and trying to add this device to support certificate based authentication.

```
CERT=$(openssl x509 -in device-4711-cert.pem -noout -pubkey | sed '/^---/d' | awk '{ printf "%s", $0 }')
I got the value “MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEtrF+3Bgyi25Mvj/AB3eB2MkOYttS0oN2n8pjZ2L8ZJN5KrO1pMRmk+TyrHT5BbRZYujYjycxqzzdICxk3Yi3pg==”.
```

I was referring to the [https://eclipse.dev/hono/docs/api/credentials/](https://eclipse.dev/hono/docs/api/credentials/) Here I got some way to do it using credentials api but it needs AMQP adapter to be available on Hono.

As I only have MQTT, Kafka and Registry-Server pods available on my dev environment Hono. Is there any way to do it using the Registry-Server while registering the device or any other easiest way to do it?

**Expected End Goal:**
mosquitto_pub -h <MQTT-Host Here> -p 8883 --cert device-4711-cert.pem --key device-4711-key.pem --cafile trusted-certs.pem -t t -m '{"temp": 5}'

This should work with provided certificate.

**Thanks in advance!**

**Comment by sophokles73** [comment]
Did you create your own certificates?

**Comment by pratikmhatre11** [comment]
I created the certificates using **create_certs.sh**. Please see below attached screenshot which where you can **create_certs.sh**.

<img width="276" alt="Screenshot 2023-12-02 at 10 26 30 PM" src="https://github.com/eclipse-hono/hono/assets/152691326/3cd5f422-f187-4c02-adc3-5126c1713d3d">

It creates several certificates under certs folder (shown in above image), from which I am using below 4 certificates.
1. device-4711-cert.pem
2. device-4711-key.pem
3. default_tenant-cert.pem
4. trusted-certs.pem

Please find content of create_certs.sh file for your reference:

```
#!/bin/bash
#*******************************************************************************
# Copyright (c) 2016, 2022 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# http://www.eclipse.org/legal/epl-2.0
#
# SPDX-License-Identifier: EPL-2.0
#*******************************************************************************

# A simple shell script for generating example certificates to be used with Hono.

DIR=certs
AUTH_SERVER_KEY_STORE=authServerKeyStore.p12
AUTH_SERVER_KEY_STORE_PWD=authkeys
CMD_ROUTER_KEY_STORE=commandRouterKeyStore.p12
CMD_ROUTER_KEY_STORE_PWD=commandrouterkeys
DEVREG_SERVER_KEY_STORE=deviceRegistryKeyStore.p12
DEVREG_SERVER_KEY_STORE_PWD=deviceregistrykeys
MQTT_ADAPTER_KEY_STORE=mqttKeyStore.p12
MQTT_ADAPTER_KEY_STORE_PWD=mqttkeys
HTTP_ADAPTER_KEY_STORE=httpKeyStore.p12
HTTP_ADAPTER_KEY_STORE_PWD=httpkeys
LORA_ADAPTER_KEY_STORE=loraKeyStore.p12
LORA_ADAPTER_KEY_STORE_PWD=lorakeys
ARTEMIS_KEY_STORE=artemisKeyStore.p12
ARTEMIS_KEY_STORE_PWD=artemiskeys
COAP_ADAPTER_KEY_STORE=coapKeyStore.p12
COAP_ADAPTER_KEY_STORE_PWD=coapkeys
AMQP_ADAPTER_KEY_STORE=amqpKeyStore.p12
AMQP_ADAPTER_KEY_STORE_PWD=amqpkeys
EXAMPLE_GATEWAY_KEY_STORE=exampleGatewayKeyStore.p12
EXAMPLE_GATEWAY_KEY_STORE_PWD=examplegatewaykeys
# set to either EC or RSA
KEY_ALG=EC

function create_key { 

  echo ""
  if [ $KEY_ALG == "EC" ]
  then
    openssl ecparam -name prime256v1 -genkey -noout | openssl pkcs8 -topk8 -nocrypt -inform PEM -outform PEM -out $DIR/$1
  else
    openssl genrsa 4096 | openssl pkcs8 -topk8 -nocrypt -inform PEM -outform PEM -out $DIR/$1
  fi
}

function create_cert {

  echo ""
  echo "creating $1 key and certificate"
  create_key $1-key.pem
  openssl req -config ca_opts -new -key $DIR/$1-key.pem -subj "/C=CA/L=Ottawa/O=Eclipse IoT/OU=Hono/CN=$1" | \
    openssl x509 -req -extfile ca_opts -extensions req_ext_$1 -out $DIR/$1.pem -days 365 -CA $DIR/ca-cert.pem -CAkey $DIR/ca-key.pem -CAcreateserial
  cat $DIR/$1.pem $DIR/ca-cert.pem > $DIR/$1-cert.pem && rm $DIR/$1.pem
  if [ $2 ]
  then
    echo "adding key/cert for $1 to key store $DIR/$2"
    openssl pkcs12 -export -inkey $DIR/$1-key.pem -in $DIR/$1-cert.pem -out $DIR/$2 -password pass:$3
  fi
}

function create_client_cert {
  echo ""
  echo "creating client key and certificate for device $1"
  create_key device-$1-key.pem
  openssl req -new -key "$DIR/device-$1-key.pem" -subj "/C=CA/L=Ottawa/O=Eclipse IoT/OU=Hono/CN=Device $1" | \
    openssl x509 -req -out "$DIR/device-$1-cert.pem" -days 365 -CA $DIR/default_tenant-cert.pem -CAkey $DIR/default_tenant-key.pem -CAcreateserial
  SUBJECT=$(openssl x509 -in "$DIR/device-$1-cert.pem" -noout -subject -nameopt RFC2253)
  echo "cert.device-$1.$SUBJECT" >> $DIR/device-certs.properties
}

if [ -d $DIR ]
then
rm $DIR/*.pem
rm $DIR/*.p12
rm $DIR/*.jks
rm $DIR/*.properties
else
mkdir $DIR
fi

echo "creating root key and certificate"
create_key root-key.pem
openssl req -x509 -config ca_opts -new -key $DIR/root-key.pem -out $DIR/root-cert.pem -days 365 -subj "/C=CA/L=Ottawa/O=Eclipse IoT/OU=Hono/CN=root"

echo ""
echo "creating CA key and certificate"
create_key ca-key.pem
openssl req -config ca_opts -reqexts intermediate_ext -new -key $DIR/ca-key.pem -days 365 -subj "/C=CA/L=Ottawa/O=Eclipse IoT/OU=Hono/CN=ca" | \
 openssl x509 -req -extfile ca_opts -extensions intermediate_ext -out $DIR/ca-cert.pem -days 365 -CA $DIR/root-cert.pem -CAkey $DIR/root-key.pem -CAcreateserial

echo ""
echo "downloading CA and root certificates from Let's Encrypt"
curl https://letsencrypt.org/certs/lets-encrypt-r3.pem > $DIR/lets-encrypt-r3.pem
curl https://letsencrypt.org/certs/isrgrootx1.pem > $DIR/isrgrootx1.pem

echo ""
echo "creating PEM trust store ($DIR/trusted-certs.pem) containing CA certificates"
cat $DIR/ca-cert.pem $DIR/root-cert.pem $DIR/lets-encrypt-r3.pem $DIR/isrgrootx1.pem > $DIR/trusted-certs.pem
rm $DIR/lets-encrypt-r3.pem $DIR/isrgrootx1.pem

echo ""
echo "creating CA key and certificate for DEFAULT_TENANT"
create_key default_tenant-key.pem
openssl req -x509 -key $DIR/default_tenant-key.pem -out $DIR/default_tenant-cert.pem -days 365 -subj "/C=CA/L=Ottawa/O=Eclipse IoT/OU=Hono/CN=DEFAULT_TENANT_CA"

echo ""
echo "extracting trust anchor information from tenant CA cert"
CA_SUBJECT=$(openssl x509 -in $DIR/default_tenant-cert.pem -noout -subject -nameopt RFC2253 | sed s/^subject=//)
PK=$(openssl x509 -in $DIR/default_tenant-cert.pem -noout -pubkey | sed '/^---/d' | awk '{ printf "%s", $0 }')
NOT_BEFORE=$(date -jf "%b %e %T %Y %Z" "$(openssl x509 -in $DIR/default_tenant-cert.pem -noout -startdate -nameopt RFC2253 | sed 's/^notBefore=//')" "+%Y-%m-%dT%H:%M:%S%z")
NOT_AFTER=$(date -jf "%b %e %T %Y %Z" "$(openssl x509 -in $DIR/default_tenant-cert.pem -noout -startdate -nameopt RFC2253 | sed 's/^notBefore=//')" "+%Y-%m-%dT%H:%M:%S%z")
echo "{" > default_tenant-trusted-ca.json
echo "  \"subject-dn\": \"$CA_SUBJECT\"," >> default_tenant-trusted-ca.json
echo "  \"public-key\": \"$PK\"," >> default_tenant-trusted-ca.json
echo "  \"algorithm\": \"$KEY_ALG\"," >> default_tenant-trusted-ca.json
echo "  \"not-before\": \"$NOT_BEFORE\"," >> default_tenant-trusted-ca.json
echo "  \"not-after\": \"$NOT_AFTER\"" >> default_tenant-trusted-ca.json
echo "}" >> default_tenant-trusted-ca.json

create_cert qdrouter
create_cert auth-server $AUTH_SERVER_KEY_STORE $AUTH_SERVER_KEY_STORE_PWD
create_cert device-registry $DEVREG_SERVER_KEY_STORE $DEVREG_SERVER_KEY_STORE_PWD
create_cert command-router $CMD_ROUTER_KEY_STORE $CMD_ROUTER_KEY_STORE_PWD
create_cert http-adapter $HTTP_ADAPTER_KEY_STORE $HTTP_ADAPTER_KEY_STORE_PWD
create_cert lora-adapter $LORA_ADAPTER_KEY_STORE $LORA_ADAPTER_KEY_STORE_PWD
create_cert mqtt-adapter $MQTT_ADAPTER_KEY_STORE $MQTT_ADAPTER_KEY_STORE_PWD
create_cert artemis $ARTEMIS_KEY_STORE $ARTEMIS_KEY_STORE_PWD
create_cert coap-adapter $COAP_ADAPTER_KEY_STORE $COAP_ADAPTER_KEY_STORE_PWD
create_cert amqp-adapter $AMQP_ADAPTER_KEY_STORE $AMQP_ADAPTER_KEY_STORE_PWD
create_cert example-gateway $EXAMPLE_GATEWAY_KEY_STORE $EXAMPLE_GATEWAY_KEY_STORE_PWD
create_cert kafka

create_client_cert 4711
```

**Comment by sophokles73** [comment]
This is most likely because you have not adapted the MQTT adapter's server certificate's alternative subject name (ASN) list to include its IP address (which you are probably using to connect to the adapter). Your client therefore cannot verify the MQTT adapter's server cert because it does not match the IP address that your mosquitto-pub client used to connect to the adapter ...

**Comment by pratikmhatre11** [comment]
**Thank you @sophokles73 for your reply.**

It would be really helpful if you can provide me some reference link where I can find more insights on how to add MQTT IP address in MQTT adapter's server certificate's alternative subject name (ASN) list. 
Is it to set somewhere while creating the certificates using create_certs.sh?

**Also a quick question from my side, Is it necessary to have AMQP adapter for certificate-base authentication?**

As I was going through the documentation I noticed on [Auth Server Configuration Page](https://eclipse.dev/hono/docs/admin-guide/auth-server-config/) that It uses AMQP host and port in OS Environment variable. Also on [Credentials API Specification Page](https://eclipse.dev/hono/docs/api/credentials/) credentials service uses AMQP to fetch the credentials. 
Can that be a issue in my case?

**Comment by sophokles73** [comment]
@pratikmhatre11 can this be closed?

**Comment by pratikmhatre11** [comment]
Ohh Sorry. I forgot to close this issue. Closing this issue.

Thanks @sophokles73 for your quick support. With your help I was able to solve the issue and now I am able to send the data on MQTT topic on my server with certificate based authentication.

You were correct. The issue was with ASN. It was me that I had less knowledge regarding the certificates. Thanks again for all your help @sophokles73.
